_posts/2019-05-29-angular-onpush-strategy.md

```diff
@@ -0,0 +1,93 @@
+
+<img alt="Sprawdzanie komponentów przez mechanizm detekcji zmian" src="/assets/img/posts/2019-05-29-angular-onpush-strategy/change-detection.png" />
+
+Strzałki obrazują kierunek przechodzenia przez drzewo mechanizmu detekcji. Jest to prosty przykład, gdyż drzewo jest bardzo małe. Wyobraźmy sobie jednak drzewo zbudowane z setek komponentów. Z każdą zmianą Angular musiałby na nowo przeszukać całe drzewo komponentów celem znalezienia tego komponentu, który wyemitował zmianę. Dość sporo obliczeń, czyż nie?
+
+## Strategia onPush
+Na szczęście Angular pozwala nam na zmianę domyślnej strategii detekcji zmian. Jeżeli nie chcemy korzystać z domyślnego mechanizmu, to na ratunek przychodzi nam strategia onPush! Strategia ta mówi nam, że komponent zależny jest tylko i wyłącznie od swoich inputów. Taki komponent nazywamy "czystym". Zmiana propagowana jest w momencie zmiany referencji inputów komponentu jak i w przypadku wyemitowania zdarzenia DOM w szablonie komponentu (np. kliknięcie w przycisk - event onclick). Co więcej, komponent emitujący zmianę z wykorzystaniem strategii onPush powiadamia mechanizm detekcji Angulara, że to właśnie on wyemitował zmianę! To drastycznie zmniejsza koszt przeszukania drzewa komponentów, gdyż Angular wie, którego komponentu szukać, albo który komponent pominąć. Zdjęcie poniżej pozwoli zobrazować tę sytuację.
+<img alt="Sprawdzanie komponentów przez mechanizm detekcji zmian" src="/assets/img/posts/2019-05-29-angular-onpush-strategy/onpush-detection.png" />
+
+### Jak używać?
+Używanie takiej strategii wymusza na nas zmiany podejścia odnośnie projektowania naszych komponentów. Inputy czystego komponentu powinny być niezmienialne, co oznacza, że wartości naszych inputów powinny być aktualizowane przez zmianę referencji, a nie wartości. Prosty przykład:
+```typescript
+@Input() cowDonation: { donation: number };
+
+onClickUpdateDonation() {
+    this.cowDonation.donation = 500;
+}
+
+```
+Powyższa zmiana nie zadziała, ponieważ zmieniamy wartość, a nie referencję. Aby strategia onPush zadziałała, wartość dotacji musimy zmienić poprzez zmianę referencji, czyli przykładowo:
+```typescript
+@Input() cowDonation: { donation: number };
+
+updateDonation() {
+    this.cowDonation = {
+        donation: 500
+    };
+}
+```
+Zdarzenia DOM są zdarzeniami asynchronicznymi, więc moglibyśmy wyciągnąć wniosek: detekcja zmian zadziała, kiedy użyjemy takich funkcji asynchronicznych jak setTimeout, setInterval albo subskrypcja do Observable'a zwracanego przez serwis HTTP, prawda? Otóż nie. Na szczęście w przypadku bytów typu Observable Angular przychodzi nam z pomocą i udostępnia AsyncPipe. Dlaczego to działa z użyciem AsyncPipe a nie z manualną subskrypcją? Zajrzyjmy więc w kod:
+```typescript
+_updateLatestValue(async, value) {
```

Comment:
imho przydałoby się pokazać sposób pozyskania this._ref

Reply:
ale z tego co rozumiem to jest wewnętrzny kodzik AsyncPipe zacytowany ;-)

Reply:
w takiej sytuacji możemy dać ew. link

Reply:
to jest fragment wyciągnięty z kodów AsyncPipe, mogę to jakoś opisać, dokumentacja akurat na ten temat nic nie mówi :P